src/DotNetWorker.Core/Diagnostics/FunctionActivitySource.cs

```diff
@@ -0,0 +1,64 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using Microsoft.Extensions.Options;
+
+namespace Microsoft.Azure.Functions.Worker.Diagnostics
+{
+    internal class FunctionActivitySourceFactory
+    {
+        private static readonly ActivitySource _activitySource = new(TraceConstants.FunctionsActivitySource, TraceConstants.FunctionsActivitySourceVersion);
+        private readonly string _schemaVersionUrl;
+        private readonly Lazy<IReadOnlyDictionary<string, string>> _attributeMap;
+
+        public FunctionActivitySourceFactory(IOptions<WorkerOptions> options)
+        {
+            _attributeMap = new Lazy<IReadOnlyDictionary<string, string>>(() => GetMapping(options.Value.OpenTelemetrySchemaVersion));
+            _schemaVersionUrl = TraceConstants.OpenTelemetrySchemaMap[options.Value.OpenTelemetrySchemaVersion];
+        }
+
+        public Activity? StartInvoke(FunctionContext context)
+        {
+            var activity = _activitySource.StartActivity(TraceConstants.FunctionsInvokeActivityName, ActivityKind.Internal, context.TraceContext.TraceParent,
+                tags: GetTags(context));
```

Comment:
To optimize the non-listener scenario, we can build and add tags only if `activity != null`:

``` CSharp
var activity = _activitySource.StartActivity(TraceConstants.FunctionsInvokeActivityName, ActivityKind.Internal, context.TraceContext.TraceParent);

if (activity is null)
{
    return null;
}

activity.TraceStateString = context.TraceContext.TraceState;

// add tags here.
return activity;
```

I also like to invert the null check and short circuit, allowing the rest of my logic to be outside of an `if` statement. But that is all personal preference.

Reply:
`GetTags()` is using `yield return` for this situation. No tags are instantiated unless the `Activity` is created. The trick is that I wanted as much info as possible sent to the ActivityStarted handler in the listener -- in order to have any tags at that point, they need to be passed into `StartActivity()`.

Reply:
omg -- 2 seconds after sending this, I see there's a `CreateActivity()` that lets you handle this in a more sane matter -- you can call that, then separately call `Start()`. This was added in 6.0.0 -- but I originally wrote all this with 5.0.0 which is why I never saw it.

Reply:
I've actually wrapped this in `_activitySource.HasListeners()` before I do anything -- this should cover it I believe. It also makes it more clear for future maintainers of this.